Use detailed errors for BMP encoder and decoder

lo48576 opened the pull request:
This is part of #1134, rework for `bmp` module.

Most of the errors were `FormatError` and can be converted simply to `DecodingError`.
However, I'm not sure how `ImageError::UnsupportedError` should be translated:

* Which should be used for "Unsupported Bitmap Header", `DecoderError` or `UnsupportedError`?
    + This is emitted when BMP header type (more precisely, its size) is unknown.
* Which should be used for "Unsupported image type", `DecoderError` or `UnsupportedError`?
    + This is emitted when the BMP image uses PNG or JPEG compression, or other unknown image types.

Additionally, if `UnsupportedError` should be used for them, what error kind (variants of `UnsupportedErrorKind`) should be used for them?

* `Color` is not appropriate for them.
* `Format` would be better, but I'm not sure the errors are really "unsupported image formats" (because the decoder is sure it is reading BMP!), and it loses precision of error information.
* `GenericFeature` can be used to translate the error naively, but it is "discouraged and is likely to get deprecated".

lo48576 commented on src/bmp/decoder.rs:
```diff
@@ -670,42 +700,46 @@ impl<R: Read + Seek> BmpDecoder<R> {
-                return Err(ImageError::UnsupportedError(
+                return Err(ImageError::Decoding(DecodingError::with_message(
+                    ImageFormat::Bmp.into(),
                     "Unsupported image type".to_string(),
```
Which should be used, `DecodingError` or `UnsupportedError`?

Possible candidate:

```rust
                return Err(ImageError::Unsupported(
                    UnsupportedError::from_format_and_kind(
                        ImageFormat::Bmp.into(),
                        UnsupportedErrorKind::GenericFeature(format!(
                            "image type {}",
                            image_type_u32
                        )),
                    ),
                ));
```

HeroicKatora replied:
This should be split into two: `UnsupportedError` for values `BI_JPEG = 4` and `BI_PNG = 5` and then a `DecodingError` for all other values.

lo48576 commented on src/bmp/decoder.rs:
```diff
@@ -770,9 +804,10 @@ impl<R: Read + Seek> BmpDecoder<R> {
-                    return Err(ImageError::UnsupportedError(
+                    return Err(ImageError::Decoding(DecodingError::with_message(
+                        ImageFormat::Bmp.into(),
                         "Unsupported Bitmap Header".to_string(),
```
Which should be used, `DecodingError` or `UnsupportedError`?

Possible candidate:

```rust
                    return Err(ImageError::Unsupported(
                        UnsupportedError::from_format_and_kind(
                            ImageFormat::Bmp.into(),
                            UnsupportedErrorKind::GenericFeature(
                                "Unsupported Bitmap Header".to_owned(),
                            ),
                        ),
                    ))
```

HeroicKatora replied:
Similar deal. There are technically more header sizes allowed corresponding to other header types which we do not precisely know (for the future and because there are unofficial extensions...). I think in this case it should always fail with `Unsupported` if the header size is larger than `BITMAPCOREHEADER_SIZE` and fail with `DecodingError` if it is smaller.

lo48576 commented:
Updated (and fixup'ed) the commit to use both `DecodingError` and `UnsupportedError`.

I noticed that [there are other compression types (0xb, 0xc, 0xd) for CMYK](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-wmf/4e588f70-bd92-4a6f-b77f-35d0feaf7a57), so I added match arms for them to emit more precise error message.

lo48576 commented:
Sorry, force-pushed [little update](https://github.com/image-rs/image/compare/c7f751f426b09edc5c61f7a88f9d3ee71c594596..b84de256b4dbe1047045b772d347d877f62bed09): added image type to the error message for unknown image type.

fintelia commented on src/bmp/decoder.rs:
```diff
@@ -670,42 +702,73 @@ impl<R: Read + Seek> BmpDecoder<R> {
+                // Unknown compression enumeration.
+                return Err(ImageError::Decoding(DecodingError::with_message(
+                    ImageFormat::Bmp.into(),
+                    format!("Unknown image type {}", image_type_u32),
```
Would it be accurate to say "unknown image compression type {}"?

lo48576 replied:
Should "Unknown" be all lowercase?

HeroicKatora replied:
It's more consistent to have an upper case at the start, also with other errors. So no :smiley:

HeroicKatora commented:
Thank you! One format down.
